File: src/api/routers/spoke/governance.py
"""Governance metrics router — /spoke/governance/metric/...

Handler naming: BACKEND.md §Route Handler Naming Convention.
Auth: authenticated; writes require Editor or Admin (require_writer).
Spec: API.md §Metric (/spoke/governance/metric).
"""

from datetime import datetime
from typing import Annotated

import httpx
from fastapi import APIRouter, Depends, Path, Query, Response, status

from src.api.auth.dependencies import AuthContext, require_authenticated, require_writer
from src.api.dependencies import get_airflow_client, get_metrics_service
from src.api.schemas.common import parse_sort
from src.api.schemas.events import EventListResponse, EventResponse
from src.api.schemas.metrics import (
    CreateMetricConfigRequest,
    MetricAttrResponse,
    MetricDefinitionListResponse,
    MetricDefinitionResponse,
    MetricResultListResponse,
    MetricResultResponse,
    MetricRunResultResponse,
    PatchMetricConfigRequest,
    ReplaceMetricConfigRequest,
)
from src.backend.metrics.service import MetricDefinitionRecord, MetricsService
from src.shared.db.models import Event, MetricDefinition, MetricResult
from src.shared.exceptions import ConflictError, NotImplementedAPIError
from src.shared.settings import settings
from src.workflows.airflow.client import AirflowClient

_METRIC_ID_PATTERN = r"^[a-z0-9][a-z0-9-]{0,62}[a-z0-9]$|^[a-z0-9]$"
MetricIdParam = Annotated[str, Path(pattern=_METRIC_ID_PATTERN)]

router = APIRouter(
    prefix="/metric",
    tags=["governance/metric"],
    dependencies=[Depends(require_authenticated)],
)


def _definition_response(m: "MetricDefinitionRecord") -> MetricDefinitionResponse:
    return MetricDefinitionResponse(
        id=m.id,
        mode=m.mode,
        is_enabled=m.is_enabled,
        metric_type=m.metric_type,
        title=m.title,
        description=m.description,
        metrics=m.metrics,
        metric_conf=m.metric_conf,
        schedule_tier=m.schedule_tier,
        dataset_filter=m.dataset_filter,
        created_at=m.created_at,
        updated_at=m.updated_at,
    )


@router.post("", response_model=MetricDefinitionResponse, status_code=status.HTTP_201_CREATED)
async def post_metric(
    body: CreateMetricConfigRequest,
    response: Response,
    service: MetricsService = Depends(get_metrics_service),
    _writer: AuthContext = Depends(require_writer),
) -> MetricDefinitionResponse:
    """Create a new metric definition; ``metric_id`` is supplied in the request body.

    Returns 409 METRIC_EXISTS when the id already exists.
    Returns 501 NOT_IMPLEMENTED when ``mode`` is 'passive'.
    """
    if body.mode == "passive":
        raise NotImplementedAPIError("Passive mode is reserved for a future release")

    metric = await service.create_metric_config(
        metric_id=body.metric_id,
        mode=body.mode,
        metric_type=body.metric_type,
        title=body.title,
        description=body.description,
        metrics=body.metrics,
        metric_conf=body.metric_conf,
        dataset_filter=body.dataset_filter,
        schedule_tier=body.schedule_tier,
        is_enabled=body.is_enabled,
    )
    response.status_code = status.HTTP_201_CREATED
    return _definition_response(metric)


@router.get("", response_model=MetricDefinitionListResponse)
async def get_metric_list(
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=20, ge=1, le=1000),
    sort: str | None = Query(default=None),
    metric_type: str | None = Query(default=None),
    mode: str | None = Query(default=None),
    is_enabled: bool | None = Query(default=None),
    service: MetricsService = Depends(get_metrics_service),
) -> MetricDefinitionListResponse:
    """List metric definitions with optional metric_type, mode, and is_enabled filters."""
    order_by = parse_sort(
        sort,
        {
            "created_at": MetricDefinition.created_at,
            "updated_at": MetricDefinition.updated_at,
            "title": MetricDefinition.title,
        },
        None,
    )
    metrics, total_count = await service.list_metrics(
        offset=offset,
        limit=limit,
        metric_type_filter=metric_type,
        mode_filter=mode,
        is_enabled_filter=is_enabled,
        order_by=order_by,
    )
    return MetricDefinitionListResponse(
        offset=offset,
        limit=limit,
        total_count=total_count,
        metrics=[_definition_response(m) for m in metrics],
    )


@router.get("/{metric_id}", response_model=MetricDefinitionResponse)
async def get_metric(
    metric_id: MetricIdParam,
    service: MetricsService = Depends(get_metrics_service),
) -> MetricDefinitionResponse:
    """Get metric summary (identity, mode, metric_type, enabled status)."""
    metric = await service.get_metric(metric_id)
    return _definition_response(metric)


@router.get("/{metric_id}/attr", response_model=MetricAttrResponse)
async def get_metric_attr(
    metric_id: MetricIdParam,
    service: MetricsService = Depends(get_metrics_service),
) -> MetricAttrResponse:
    """Get metric attributes overview (mode, metric_type, schedule_tier, enabled, latest values)."""
    attr = await service.get_metric_attr(metric_id)
    return MetricAttrResponse(**attr)


@router.get("/{metric_id}/attr/conf", response_model=MetricDefinitionResponse)
async def get_metric_conf(
    metric_id: MetricIdParam,
    service: MetricsService = Depends(get_metrics_service),
) -> MetricDefinitionResponse:
    """Get full metric definition."""
    metric = await service.get_metric_config(metric_id)
    return _definition_response(metric)


@router.put("/{metric_id}/attr/conf", response_model=MetricDefinitionResponse)
async def put_metric_conf(
    metric_id: MetricIdParam,
    body: ReplaceMetricConfigRequest,
    service: MetricsService = Depends(get_metrics_service),
    _writer: AuthContext = Depends(require_writer),
) -> MetricDefinitionResponse:
    """Replace an existing metric definition.

    Returns 404 METRIC_NOT_FOUND when the id is absent (use POST /spoke/governance/metric
    to create).
    Returns 501 NOT_IMPLEMENTED when mode is 'passive'.
    """
    if body.mode == "passive":
        raise NotImplementedAPIError("Passive mode is reserved for a future release")

    metric = await service.replace_metric_config(
        metric_id=metric_id,
        mode=body.mode,
        metric_type=body.metric_type,
        title=body.title,
        description=body.description,
        metrics=body.metrics,
        metric_conf=body.metric_conf,
        dataset_filter=body.dataset_filter,
        schedule_tier=body.schedule_tier,
        is_enabled=body.is_enabled,
    )
    return _definition_response(metric)


@router.patch("/{metric_id}/attr/conf", response_model=MetricDefinitionResponse)
async def patch_metric_conf(
    metric_id: MetricIdParam,
    body: PatchMetricConfigRequest,
    service: MetricsService = Depends(get_metrics_service),
    _writer: AuthContext = Depends(require_writer),
) -> MetricDefinitionResponse:
    """Update metric definition fields.

    Returns 501 NOT_IMPLEMENTED when the patch sets mode to 'passive'.
    """
    if body.mode == "passive":
        raise NotImplementedAPIError("Passive mode is reserved for a future release")

    patch = body.model_dump(exclude_unset=True)
    metric = await service.patch_metric_config(metric_id, patch)
    return _definition_response(metric)


@router.delete("/{metric_id}/attr/conf", status_code=status.HTTP_204_NO_CONTENT)
async def delete_metric_conf(
    metric_id: MetricIdParam,
    service: MetricsService = Depends(get_metrics_service),
    _writer: AuthContext = Depends(require_writer),
) -> None:
    """Remove a metric definition and its configuration."""
    await service.delete_metric_config(metric_id)


@router.get("/{metric_id}/attr/result", response_model=MetricResultListResponse)
async def get_metric_results(
    metric_id: MetricIdParam,
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=20, ge=1, le=1000),
    sort: str | None = Query(default=None),
    from_time: datetime | None = Query(default=None, alias="from"),
    to_time: datetime | None = Query(default=None, alias="to"),
    service: MetricsService = Depends(get_metrics_service),
) -> MetricResultListResponse:
    """Get measurement results (dict-valued timeseries; ?from/to for time range)."""
    order_by = parse_sort(sort, {"measured_at": MetricResult.measured_at}, None)
    results, total_count = await service.get_results(
        metric_id,
        from_dt=from_time,
        to_dt=to_time,
        offset=offset,
        limit=limit,
        order_by=order_by,
    )
    return MetricResultListResponse(
        offset=offset,
        limit=limit,
        total_count=total_count,
        results=[
            MetricResultResponse(
                id=r.id,
                metric_id=r.metric_id,
                values=r.values,
                breakdown=r.breakdown,
                measured_at=r.measured_at,
            )
            for r in results
        ],
    )


@router.post("/{metric_id}/method/run", response_model=MetricRunResultResponse)
async def post_metric_run(
    metric_id: MetricIdParam,
    dry_run: bool = Query(default=False),
    airflow: AirflowClient = Depends(get_airflow_client),
    service: MetricsService = Depends(get_metrics_service),
    _writer: AuthContext = Depends(require_writer),
) -> MetricRunResultResponse:
    """Trigger a metric measurement run.

    Pass ``?dry_run=true`` to simulate the measurement without persisting results.
    Concurrent runs return 409 METRIC_RUNNING.
    Returns 409 METRIC_DISABLED when the metric is disabled and ``dry_run`` is not true.
    """
    definition = await service.get_metric(metric_id)
    if not definition.is_enabled and not dry_run:
        raise ConflictError(
            "METRIC_DISABLED",
            f"Metric {metric_id} is disabled; only dry-run is permitted",
        )

    workflow_id = f"metrics-{metric_id}"
    await airflow.check_no_duplicate("metrics", "workflow_id", workflow_id, "METRIC_RUNNING")
    try:
        dag_run = await airflow.trigger_and_wait(
            "metrics",
            conf={
                "callback_base_url": settings.airflow_callback_base_url,
                "metric_id": metric_id,
                "dry_run": str(dry_run).lower(),
                "workflow_id": workflow_id,
            },
        )
    except httpx.HTTPStatusError as exc:
        # check_no_duplicate is non-atomic; concurrent callers race past it and
        # Airflow rejects all but one with 409 "DAG run already exists".
        if exc.response.status_code == 409:
            raise ConflictError(
                "METRIC_RUNNING",
                f"Metric measurement is already running for {metric_id}",
            ) from exc
        raise
    xcom_value = await airflow.fetch_task_xcom(
        dag_id="metrics",
        dag_run_id=dag_run.dag_run_id,
        task_id="run_metric",
    )
    if (
        isinstance(xcom_value, dict)
        and xcom_value.get("status") == "error"
        and xcom_value.get("detail", {}).get("error_code") == "METRIC_RUNNING"
    ):
        raw_msg = xcom_value["detail"].get(
            "message", f"Metric measurement is already running for {metric_id}"
        )
        msg = (raw_msg[:200] + "…") if len(raw_msg) > 200 else raw_msg
        raise ConflictError("METRIC_RUNNING", msg)
    if isinstance(xcom_value, dict):
        return MetricRunResultResponse(
            run_id=xcom_value.get("run_id", dag_run.dag_run_id),
            status=xcom_value.get("status", dag_run.state.value),
            detail=xcom_value.get("detail", {}),
        )
    return MetricRunResultResponse(
        run_id=dag_run.dag_run_id,
        status=dag_run.state.value,
        detail={},
    )


@router.get("/{metric_id}/event", response_model=EventListResponse)
async def get_metric_events(
    metric_id: MetricIdParam,
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=20, ge=1, le=1000),
    sort: str | None = Query(default=None),
    from_time: datetime | None = Query(default=None, alias="from"),
    to_time: datetime | None = Query(default=None, alias="to"),
    service: MetricsService = Depends(get_metrics_service),
) -> EventListResponse:
    """Metric run events (run completions, definition changes)."""
    order_by = parse_sort(sort, {"occurred_at": Event.occurred_at}, None)
    events, total_count = await service.get_events(
        metric_id,
        offset=offset,
        limit=limit,
        from_dt=from_time,
        to_dt=to_time,
        order_by=order_by,
    )
    return EventListResponse(
        offset=offset,
        limit=limit,
        total_count=total_count,
        events=[
            EventResponse(
                id=str(e["id"]),
                entity_type=e["entity_type"],
                entity_id=e["entity_id"],
                event_type=e["event_type"],
                status=e["status"],
                detail=e.get("detail", {}),
                occurred_at=e["occurred_at"],
            )
            for e in events
        ],
    )

File: src/frontend/lib/api/governance.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiFetch } from "@/lib/api/client";
import { usePoll } from "@/lib/hooks/use-poll";
import type {
  CreateMetricFormValues,
  DatasetFilter,
  MetricDefinition,
  MetricDefinitionListResponse,
  MetricEventListResponse,
  MetricFormValues,
  MetricResultListResponse,
  MetricRunResult,
  ScheduleTier,
} from "@/types/governance";

// ── List metrics ───────────────────────────────────────────────────────────────

interface ListMetricsParams {
  offset?: number;
  limit?: number;
  metric_type?: string;
  mode?: string;
  is_enabled?: boolean;
  sort?: string;
}

function buildListMetricsUrl(params: ListMetricsParams): string {
  const sp = new URLSearchParams();
  if (params.offset !== undefined) sp.set("offset", String(params.offset));
  if (params.limit !== undefined) sp.set("limit", String(params.limit));
  if (params.metric_type) sp.set("metric_type", params.metric_type);
  if (params.mode) sp.set("mode", params.mode);
  if (params.is_enabled !== undefined) sp.set("is_enabled", String(params.is_enabled));
  if (params.sort) sp.set("sort", params.sort);
  const qs = sp.toString();
  return `/spoke/governance/metric${qs ? `?${qs}` : ""}`;
}

export function useGovernanceMetrics(params: ListMetricsParams = {}) {
  return useQuery<MetricDefinitionListResponse>({
    queryKey: ["governance", "metrics", params],
    queryFn: () => apiFetch<MetricDefinitionListResponse>(buildListMetricsUrl(params)),
    meta: { handledInline: true },
  });
}

/** Polls enabled metrics for use on the dashboard. */
export function useEnabledMetrics() {
  return usePoll<MetricDefinitionListResponse>({
    queryKey: ["governance", "metrics", { is_enabled: true, limit: 100 }],
    queryFn: () =>
      apiFetch<MetricDefinitionListResponse>(
        buildListMetricsUrl({ is_enabled: true, limit: 100 }),
      ),
    meta: { handledInline: true },
  });
}

// ── Single metric conf ─────────────────────────────────────────────────────────

export function useMetricConf(metricId: string) {
  return useQuery<MetricDefinition>({
    queryKey: ["governance", "metrics", metricId, "conf"],
    queryFn: () => apiFetch<MetricDefinition>(`/spoke/governance/metric/${metricId}/attr/conf`),
    enabled: !!metricId,
    meta: { handledInline: true },
  });
}

// ── Metric results (timeseries) ────────────────────────────────────────────────

interface MetricResultsParams {
  from?: string;
  to?: string;
  limit?: number;
  offset?: number;
  sort?: string;
}

function buildResultsUrl(metricId: string, params: MetricResultsParams): string {
  const sp = new URLSearchParams();
  if (params.from) sp.set("from", params.from);
  if (params.to) sp.set("to", params.to);
  if (params.limit !== undefined) sp.set("limit", String(params.limit));
  if (params.offset !== undefined) sp.set("offset", String(params.offset));
  if (params.sort) sp.set("sort", params.sort);
  const qs = sp.toString();
  return `/spoke/governance/metric/${metricId}/attr/result${qs ? `?${qs}` : ""}`;
}

export function useMetricResults(metricId: string, params: MetricResultsParams = {}) {
  return usePoll<MetricResultListResponse>({
    queryKey: ["governance", "metrics", metricId, "results", params],
    queryFn: () => apiFetch<MetricResultListResponse>(buildResultsUrl(metricId, params)),
    enabled: !!metricId,
    meta: { handledInline: true },
  });
}

export function useLatestMetricResult(metricId: string) {
  return useQuery<MetricResultListResponse>({
    queryKey: ["governance", "metrics", metricId, "results", "latest"],
    queryFn: () =>
      apiFetch<MetricResultListResponse>(buildResultsUrl(metricId, { limit: 1 })),
    enabled: !!metricId,
    meta: { handledInline: true },
  });
}

// ── Metric events ──────────────────────────────────────────────────────────────

interface MetricEventParams {
  from?: string;
  to?: string;
  limit?: number;
  sort?: string;
}

function buildMetricEventUrl(metricId: string, params: MetricEventParams): string {
  const sp = new URLSearchParams();
  if (params.from) sp.set("from", params.from);
  if (params.to) sp.set("to", params.to);
  if (params.limit !== undefined) sp.set("limit", String(params.limit));
  if (params.sort) sp.set("sort", params.sort);
  const qs = sp.toString();
  return `/spoke/governance/metric/${metricId}/event${qs ? `?${qs}` : ""}`;
}

export function useMetricEvents(metricId: string, params: MetricEventParams = {}) {
  return usePoll<MetricEventListResponse>({
    queryKey: ["governance", "metrics", metricId, "events", params],
    queryFn: () =>
      apiFetch<MetricEventListResponse>(buildMetricEventUrl(metricId, params)),
    enabled: !!metricId,
    meta: { handledInline: true },
  });
}

// ── Create metric ──────────────────────────────────────────────────────────────

interface CreateMetricBody {
  metric_id: string;
  mode: string;
  is_enabled: boolean;
  metric_type: string;
  title: string;
  description: string;
  metrics: string[];
  metric_conf: Record<string, unknown>;
  schedule_tier: ScheduleTier | null;
  dataset_filter: DatasetFilter;
}

function toCreateBody(values: CreateMetricFormValues): CreateMetricBody {
  return {
    metric_id: values.metric_id,
    mode: values.mode,
    is_enabled: values.is_enabled,
    metric_type: values.metric_type,
    title: values.title,
    description: values.description,
    metrics: values.metrics,
    metric_conf: values.metric_conf,
    schedule_tier: values.schedule_tier,
    dataset_filter: values.dataset_filter,
  };
}

export function useCreateMetric() {
  const qc = useQueryClient();
  return useMutation<MetricDefinition, Error, CreateMetricFormValues>({
    mutationFn: (values) =>
      apiFetch<MetricDefinition>("/spoke/governance/metric", {
        method: "POST",
        body: JSON.stringify(toCreateBody(values)),
      }),
    meta: { handledInline: true },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: ["governance", "metrics"] });
    },
  });
}

// ── Update metric conf (PUT — full replace) ────────────────────────────────────

interface ReplaceMetricBody {
  mode: string;
  is_enabled: boolean;
  metric_type: string;
  title: string;
  description: string;
  metrics: string[];
  metric_conf: Record<string, unknown>;
  schedule_tier: ScheduleTier | null;
  dataset_filter: DatasetFilter;
}

function toReplaceBody(values: MetricFormValues): ReplaceMetricBody {
  return {
    mode: values.mode,
    is_enabled: values.is_enabled,
    metric_type: values.metric_type,
    title: values.title,
    description: values.description,
    metrics: values.metrics,
    metric_conf: values.metric_conf,
    schedule_tier: values.schedule_tier,
    dataset_filter: values.dataset_filter,
  };
}

export function useReplaceMetricConf() {
  const qc = useQueryClient();
  return useMutation<MetricDefinition, Error, { metricId: string; values: MetricFormValues }>({
    mutationFn: ({ metricId, values }) =>
      apiFetch<MetricDefinition>(`/spoke/governance/metric/${metricId}/attr/conf`, {
        method: "PUT",
        body: JSON.stringify(toReplaceBody(values)),
      }),
    meta: { handledInline: true },
    onSuccess: (_data, vars) => {
      void qc.invalidateQueries({ queryKey: ["governance", "metrics"] });
      void qc.invalidateQueries({
        queryKey: ["governance", "metrics", vars.metricId, "conf"],
      });
    },
  });
}

// ── Update metric conf (PATCH — partial) ───────────────────────────────────────

/** Partial conf body — a subset of the editable conf definition fields. */
interface PatchMetricBody {
  mode?: string;
  is_enabled?: boolean;
  metric_type?: string;
  title?: string;
  description?: string;
  metrics?: string[];
  metric_conf?: Record<string, unknown>;
  schedule_tier?: ScheduleTier | null;
  dataset_filter?: DatasetFilter;
}

export function useUpdateMetricConf() {
  const qc = useQueryClient();
  return useMutation<MetricDefinition, Error, { metricId: string; patch: PatchMetricBody }>({
    mutationFn: ({ metricId, patch }) =>
      apiFetch<MetricDefinition>(`/spoke/governance/metric/${metricId}/attr/conf`, {
        method: "PATCH",
        body: JSON.stringify(patch),
      }),
    meta: { handledInline: true },
    onSuccess: (_data, vars) => {
      void qc.invalidateQueries({ queryKey: ["governance", "metrics"] });
      void qc.invalidateQueries({
        queryKey: ["governance", "metrics", vars.metricId, "conf"],
      });
    },
  });
}

// ── Delete metric conf ─────────────────────────────────────────────────────────

export function useDeleteMetric() {
  const qc = useQueryClient();
  return useMutation<void, Error, { metricId: string }>({
    mutationFn: ({ metricId }) =>
      apiFetch<void>(`/spoke/governance/metric/${metricId}/attr/conf`, {
        method: "DELETE",
      }),
    meta: { handledInline: true },
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: ["governance", "metrics"] });
    },
  });
}

// ── Run metric ─────────────────────────────────────────────────────────────────

export function useRunMetric() {
  const qc = useQueryClient();
  return useMutation<MetricRunResult, Error, { metricId: string; dry_run?: boolean }>({
    mutationFn: ({ metricId, dry_run = false }) => {
      const url = `/spoke/governance/metric/${metricId}/method/run${dry_run ? "?dry_run=true" : ""}`;
      return apiFetch<MetricRunResult>(url, { method: "POST" });
    },
    meta: { handledInline: true },
    onSuccess: (_data, vars) => {
      void qc.invalidateQueries({
        queryKey: ["governance", "metrics", vars.metricId, "results"],
      });
      void qc.invalidateQueries({
        queryKey: ["governance", "metrics", vars.metricId, "events"],
      });
    },
  });
}
